galley/pkg/config/analysis/analyzers/authz/authorizationpolicies.go

```diff
@@ -0,0 +1,288 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package authz
+
+import (
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+	k8s_labels "k8s.io/apimachinery/pkg/labels"
+
+	"istio.io/api/annotation"
+	"istio.io/api/mesh/v1alpha1"
+	"istio.io/api/networking/v1alpha3"
+	"istio.io/api/security/v1beta1"
+
+	"istio.io/istio/galley/pkg/config/analysis"
+	"istio.io/istio/galley/pkg/config/analysis/analyzers/util"
+	"istio.io/istio/galley/pkg/config/analysis/msg"
+
+	"istio.io/istio/pkg/config/resource"
+	"istio.io/istio/pkg/config/schema/collection"
+	"istio.io/istio/pkg/config/schema/collections"
+)
+
+// AuthorizationPoliciesAnalyzer checks the validity of authorization policies
+type AuthorizationPoliciesAnalyzer struct{}
+
+var _ analysis.Analyzer = &AuthorizationPoliciesAnalyzer{}
+var meshConfig *v1alpha1.MeshConfig
+
+func (a *AuthorizationPoliciesAnalyzer) Metadata() analysis.Metadata {
+	return analysis.Metadata{
+		Name:        "auth.AuthorizationPoliciesAnalyzer",
+		Description: "Checks the validity of authorization policies",
+		Inputs: collection.Names{
+			collections.IstioMeshV1Alpha1MeshConfig.Name(),
+			collections.IstioNetworkingV1Alpha3Serviceentries.Name(),
+			collections.IstioSecurityV1Beta1Authorizationpolicies.Name(),
+			collections.K8SCoreV1Namespaces.Name(),
+			collections.K8SCoreV1Pods.Name(),
+			collections.K8SCoreV1Services.Name(),
+		},
+	}
+}
+
+func (a *AuthorizationPoliciesAnalyzer) Analyze(c analysis.Context) {
+	serviceEntryHosts := util.InitServiceEntryHostMap(c)
+	podLabelsMap := initPodLabelsMap(c)
+
+	c.ForEach(collections.IstioSecurityV1Beta1Authorizationpolicies.Name(), func(r *resource.Instance) bool {
+		a.analyzeNoMatchingWorkloads(r, c, podLabelsMap)
+		a.analyzeNamespaceNotFound(r, c)
+		a.analyzeHostNotFound(r, c, serviceEntryHosts)
```

Comment:
I'm a bit worried the host analyze could be easily producing a lot false-positive cases. The host is essentially used to match against the `Host` header in the HTTP request, which means it technically could be anything as long as istio knows how to route the request, and I think in some cases those hosts names are not known in service entry or k8s service ahead of time, meaning the analyzer will always think it's not found?

Reply:
I've just removed the analysis for that end to reduce the false positive. Let's be conservative and maybe this can be discussed in a subsequent issue if community feel it is useful to have.